### DISCUSSION TITLE
How to type async higher-order resolver in Typescript?

### DISCUSSION THREAD

**Discussion by Lee-Minhoon** [mswjs/msw#2223] [Q&A] [answered]
I referred to the "https://mswjs.io/docs/best-practices/typescript/#higher-order-response-resolvers page and wrote the code as follows.

```ts
import {
  DefaultBodyType,
  HttpResponse,
  HttpResponseResolver,
  PathParams,
} from "msw";

type HigherOrderResolver = <
  Params extends PathParams,
  RequestBodyType extends DefaultBodyType,
  ResponseBodyType extends DefaultBodyType ,
>(
  resolver: HttpResponseResolver<Params, RequestBodyType, ResponseBodyType>
) => HttpResponseResolver<Params, RequestBodyType, ResponseBodyType>;

export const withAuth: HigherOrderResolver = (resolver) => {
  return (input) => {
    const { request } = input;

    if (!request.headers.get("Authorization")) {
      return HttpResponse.json(
        { data: null, message: "Unauthorized" },
        { status: 401 }
      );
    }

    return resolver(input);
  };
};
```

However, a TypeScript error appears where the response is returned.

```
Argument of type '{ data: null; message: string; }' is not assignable to parameter of type 'NoInfer<ResponseBodyType>'.
  Type '{ data: null; message: string; }' is not assignable to type 'ResponseBodyType'.
    '{ data: null; message: string; }' is assignable to the constraint of type 'ResponseBodyType', but 'ResponseBodyType' could be instantiated with a different subtype of constraint 'DefaultBodyType'.ts(2345)
```

This is probably due to the application of the 'NoInfer' type.

Injecting the exact type into the generic makes it difficult to write a general-purpose type.

Is there a good way to solve this?

**Comment by trbagley-gresham** [comment]
I'd also be interested to know if there's a good solution for this. I'm trying to replicate the [auth example given for plain JavaScript](https://mswjs.io/docs/recipes/higher-order-resolver) but with TypeScript.

**Comment by grgmo** [comment]
was getting the same error, changed it to throw the error instead which stopped `ts` error, followed this example https://mswjs.io/docs/basics/mocking-responses/#throwing-responses

```
throw HttpResponse.json(
        { data: null, message: "Unauthorized" },
        { status: 401 }
      );
```

**Comment by kettanaito** [comment, marked as answer]
Hi. We've simplified the annotations for higher-order handlers/resolvers since this question has been posted. Please use the following types:

- `HttpHandler`: https://github.com/mswjs/msw/blob/3cc9d9f0773cb41c3256897426ca0a1523ce90a8/src/core/index.ts#L8
- `HttpResponseResolver`: https://github.com/mswjs/msw/blob/3cc9d9f0773cb41c3256897426ca0a1523ce90a8/src/core/index.ts#L46

You can find the same types for GraphQL.
